sc-memory/sc-core/sc-store/sc-fs-storage/sc_fs_storage.c

```diff
@@ -0,0 +1,382 @@
+/*
+ * This source file is part of an OSTIS project. For the latest info, see http://ostis.net
+ * Distributed under the MIT License
+ * (See accompanying file COPYING.MIT or copy at http://opensource.org/licenses/MIT)
+ */
+
+#include <glib.h>
+#include <glib/gstdio.h>
+
+#include "sc_fs_storage.h"
+
+#ifdef SC_DICTIONARY_FS_STORAGE
+#  include "sc_dictionary_fs_storage.h"
+#elif SC_ROCKSDB_FS_STORAGE
+#  include "sc_rocksdb_fs_storage.h"
+#  include "../sc_link_helpers.h"
+#endif
+
+#include "sc_file_system.h"
+
+#include "../sc_segment.h"
+#include "../../sc_memory_version.h"
+
+sc_char * repo_path = 0;
+sc_char segments_path[MAX_PATH_LENGTH];
+
+int _checksum_type()
+{
+  return G_CHECKSUM_SHA512;
+}
+
+sc_uint8 _checksum_get_size()
+{
+  return (sc_uint8)g_checksum_type_get_length(_checksum_type());
+}
+
+// ----------------------------------------------
+
+sc_bool sc_fs_storage_initialize(const sc_char * path, sc_bool clear)
+{
+  g_message("Initialize sc-storage from path: %s", path);
+  g_snprintf(segments_path, MAX_PATH_LENGTH, "%s/segments.scdb", path);
+
+  repo_path = g_strdup(path);
+
+#ifdef SC_DICTIONARY_FS_STORAGE
+  sc_dictionary_fs_storage_initialize(repo_path);
+#elif SC_ROCKSDB_FS_STORAGE
+  sc_rocksdb_fs_storage_initialize(repo_path);
+#endif
+
+  // clear repository if it needs
+  if (clear == SC_TRUE)
+  {
+    g_message("Clear memory");
+    if (g_file_test(segments_path, G_FILE_TEST_IS_REGULAR) && g_remove(segments_path) != 0)
+      g_error("Can't remove segments file: %s", segments_path);
+
+    g_message("Clear file memory");
+#ifdef SC_ROCKSDB_FS_STORAGE
+    if (sc_rocksdb_fs_storage_clear() != SC_RESULT_OK)
+    {
+      g_critical("Can't clear file memory");
+      return SC_FALSE;
+    }
+#endif
+  }
+
+  return SC_TRUE;
+}
+
+sc_bool sc_fs_storage_shutdown(sc_segment ** segments, sc_bool save_segments)
+{
+  if (save_segments == SC_TRUE)
+  {
+    g_message("Write segments");
+    sc_fs_storage_save(segments);
+  }
+
+#ifdef SC_DICTIONARY_FS_STORAGE
+  sc_dictionary_fs_storage_shutdown();
+#elif SC_ROCKSDB_FS_STORAGE
+  sc_rocksdb_fs_storage_shutdown();
+#endif
+
+  g_free(repo_path);
+  g_message("Shutdown sc-fs-storage");
+
+  return SC_FALSE;
+}
+
+void sc_fs_storage_append_sc_link(sc_element * element, sc_addr addr, sc_char * sc_string, sc_uint32 size)
+{
+#ifdef SC_DICTIONARY_FS_STORAGE
+  sc_dictionary_fs_storage_append_sc_link(addr, sc_string, size);
+#elif SC_ROCKSDB_FS_STORAGE
```

Comment:
I think that such way is more readable that proposable, because for structure using additional defines are needed but not in file and additional methods are needed for strings hashing before sc-rockdsb methods calls. May be I leave issue and it will be added in another PR after PR with strings hashing.